\documentclass[final]{beamer}\usepackage[]{graphicx}\usepackage[]{color}
%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother

\definecolor{fgcolor}{rgb}{0.345, 0.345, 0.345}
\newcommand{\hlnum}[1]{\textcolor[rgb]{0.686,0.059,0.569}{#1}}%
\newcommand{\hlstr}[1]{\textcolor[rgb]{0.192,0.494,0.8}{#1}}%
\newcommand{\hlcom}[1]{\textcolor[rgb]{0.678,0.584,0.686}{\textit{#1}}}%
\newcommand{\hlopt}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlstd}[1]{\textcolor[rgb]{0.345,0.345,0.345}{#1}}%
\newcommand{\hlkwa}[1]{\textcolor[rgb]{0.161,0.373,0.58}{\textbf{#1}}}%
\newcommand{\hlkwb}[1]{\textcolor[rgb]{0.69,0.353,0.396}{#1}}%
\newcommand{\hlkwc}[1]{\textcolor[rgb]{0.333,0.667,0.333}{#1}}%
\newcommand{\hlkwd}[1]{\textcolor[rgb]{0.737,0.353,0.396}{\textbf{#1}}}%

\usepackage{framed}
\makeatletter
\newenvironment{kframe}{%
 \def\at@end@of@kframe{}%
 \ifinner\ifhmode%
  \def\at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi%
 \def\FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {\par\unskip\endMakeFramed%
 \at@end@of@kframe}
\makeatother

\definecolor{shadecolor}{rgb}{.97, .97, .97}
\definecolor{messagecolor}{rgb}{0, 0, 0}
\definecolor{warningcolor}{rgb}{1, 0, 1}
\definecolor{errorcolor}{rgb}{1, 0, 0}
\newenvironment{knitrout}{}{} % an empty environment to be redefined in TeX

\usepackage{alltt}
\usepackage{grffile}
\mode<presentation>{\usetheme{CambridgeUSPOL}}

\usepackage[utf8]{inputenc}
\usepackage{amsfonts}
\usepackage{amsmath}
\usepackage{natbib}
\usepackage{graphicx}
\usepackage{array,booktabs,tabularx}
\newcolumntype{Z}{>{\centering\arraybackslash}X}

% rysunki
\usepackage{tikz}
\usepackage{ifthen}
\usepackage{xxcolor}
\usetikzlibrary{arrows}
\usetikzlibrary[topaths]
\usetikzlibrary{decorations.pathreplacing}
%\usepackage{times}\usefonttheme{professionalfonts}  % times is obsolete
\usefonttheme[onlymath]{serif}
\boldmath
\usepackage[orientation=portrait,size=a0,scale=1.4,debug]{beamerposter}                       % e.g. for DIN-A0 poster
%\usepackage[orientation=portrait,size=a1,scale=1.4,grid,debug]{beamerposter}                  % e.g. for DIN-A1 poster, with optional grid and debug output
%\usepackage[size=custom,width=200,height=120,scale=2,debug]{beamerposter}                     % e.g. for custom size poster
%\usepackage[orientation=portrait,size=a0,scale=1.0,printer=rwth-glossy-uv.df]{beamerposter}   % e.g. for DIN-A0 poster with rwth-glossy-uv printer check
% ...
%

\usecolortheme{seagull}
\useinnertheme{rectangles}
\setbeamercolor{item projected}{bg=darkred}
% \setbeamertemplate{enumerate items}[default]
\setbeamertemplate{caption}{\insertcaption} 
\setbeamertemplate{navigation symbols}{}
\setbeamercovered{transparent}
\setbeamercolor{block title}{fg=darkred}
\setbeamercolor{local structure}{fg=darkred}

\setbeamercolor*{enumerate item}{fg=darkred}
\setbeamercolor*{enumerate subitem}{fg=darkred}
\setbeamercolor*{enumerate subsubitem}{fg=darkred}

\setbeamercolor*{itemize item}{fg=darkred}
\setbeamercolor*{itemize subitem}{fg=darkred}
\setbeamercolor*{itemize subsubitem}{fg=darkred}

\newlength{\columnheight}
\setlength{\columnheight}{93cm}
\renewcommand{\thetable}{}
\def\andname{,}
\authornote{}

\renewcommand{\APACrefatitle}[2]{}
\renewcommand{\bibliographytypesize}{\footnotesize} 
\renewcommand{\APACrefYearMonthDay}[3]{%
  {\BBOP}{#1}
  {\BBCP}
}
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\begin{document}






\date{}
\author{Piotr Sobczyk\inst{1}*, Micha\l{}  Burdukiewicz\inst{2}, Chris Lauber\inst{3}, Pawe\l{} Mackiewicz\inst{2} \\
*Piotr.Sobczyk@pwr.edu.pl}


\institute{\small{

\textsuperscript{1}Wroc\l{}aw University of Technology, Department of Mathematics, Poland

\vspace{0.3cm}

\textsuperscript{2}University of Wroc\l{}aw, Department of Genomics, Poland 

\vspace{0.3cm}

\textsuperscript{3}Dresden University of Technology, Institute of Medical Informatics and Biometry, Poland 

}}
\title{\huge Quick Permutation Test: feature filtering of n-gram data}

\begin{frame}
  \begin{columns}
    \begin{column}{.47\textwidth}
      \begin{beamercolorbox}[center,wd=\textwidth]{postercolumn}
        \begin{minipage}[T]{.95\textwidth}
          \parbox[t][\columnheight]{\textwidth}
            {
    
        
    \begin{block}{Introduction}
N-grams (k-tuples) are vectors of n characters derived from input sequence(s). They may form continuous sub-sequences or be discontinuous. 
Important n-gram parameter is its position. Instead of just counting n-grams, one may want to count how many n-grams occur at a given position in multiple (e.g. related) sequences.

Originally developed for natural language processing, n-grams are also used in genomics~\citep{fang2011}, transcriptomics~\citep{wang2014} and proteomics~\citep{guo2014}.

\small{
       \begin{columns}[c] % the "c" option specifies center vertical alignment
    \column{.5\textwidth} 
% latex table generated in R 3.1.3 by xtable 1.7-4 package
% Fri Apr 10 11:01:39 2015
\begin{table}[ht]
\centering
\begin{tabular}{rrrrrrr}
  \hline
 & P1 & P2 & P3 & P4 & P5 & P6 \\ 
  \hline
S1 & 1 & 1 & 1 & 2 & 4 & 1 \\ 
  S2 & 2 & 3 & 1 & 2 & 1 & 2 \\ 
  S3 & 2 & 3 & 4 & 4 & 4 & 3 \\ 
   \hline
\end{tabular}
\caption{Sample sequences.  S - sequence, P - postion.} 
\end{table}

      
      
     % column designated by a command

    \column{.5\textwidth}
    
% latex table generated in R 3.1.3 by xtable 1.7-4 package
% Fri Apr 10 11:01:39 2015
\begin{table}[ht]
\centering
\begin{tabular}{rrrrr}
  \hline
 & 1 & 2 & 3 & 4 \\ 
  \hline
S1 & 4 & 1 & 0 & 1 \\ 
  S2 & 2 & 3 & 1 & 0 \\ 
  S3 & 0 & 1 & 2 & 3 \\ 
   \hline
\end{tabular}
\caption{Unigram counts.} 
\end{table}


    \end{columns}

% latex table generated in R 3.1.3 by xtable 1.7-4 package
% Fri Apr 10 11:01:39 2015
\begin{table}[ht]
\centering
\begin{tabular}{rrrrrrrrrrrrrr}
  \hline
 & P1\_1 & P2\_1 & P3\_1 & P4\_1 & P5\_1 & P6\_1 & P1\_2 & P2\_2 & P3\_2 & P4\_2 & P5\_2 & P6\_2 & P1\_3 \\ 
  \hline
S1 & 1 & 1 & 1 & 0 & 0 & 1 & 0 & 0 & 0 & 1 & 0 & 0 & 0 \\ 
  S2 & 0 & 0 & 1 & 0 & 1 & 0 & 1 & 0 & 0 & 1 & 0 & 1 & 0 \\ 
  S3 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 0 & 0 \\ 
   \hline
\end{tabular}
\caption{A fraction of possible unigrams with position information.} 
\end{table}

}    
    \end{block}
    
    \vfill
    
    \begin{block}{Curse of dimensionality}
    
Even when we limit ourselves to only continuous positioned n-grams build on $m$ possible
characters, feature space growths rapidly with the number of elements in n-gram
($n$) and the length of the sequence ($L$).    
    
The number of possible positioned n-grams: 

\begin{center}
\scalebox{0.85}{
$
n_{\text{max}} = L \times m^n
$
}

\\


\scalebox{0.91}{  
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}

{\centering \includegraphics[width=\maxwidth]{figure/unnamed-chunk-4-1} 

}



\end{knitrout}
}
\end{center}
    \end{block}
    \vfill
    
    
    \begin{block}{Feature selecting permutation tests}
    Model and statistic independent permutation tests can be used to filter features obtained through counting n-grams.
    
    During a permutation test class labels are randomly exchanged during computation of a significance statistic. p-values are defined as:
    
\begin{center}
\scalebox{0.85}{
$      
\textnormal{p-value} = \frac{N_{T_P > T_R}}{N}
$
}
\end{center}

where $N_{T_P > T_R}$ is number of times when $T_P$ (permuted test statistic) was more extreme than $T_R$ (test statistic for non-permuted data).

Permutation tests are computationally expensive (especially considering precise estimation of small p-values, because the number of permutations is inversely proportional to the interval between p-values).
      
    \end{block}
    \vfill
    
    
\begin{block}{QuiPT concept}

In each permutation, for every observation, there are four possible results:

\begin{center}
\scalebox{0.85}{
$P(Target, Feature) = (1,1)) = p \cdot q$
}
\end{center}

\\

\begin{center}
\scalebox{0.85}{
$P(Target, Feature) = (1,0)) = p \cdot (1-q)$
}
\end{center}

\\

\begin{center}
\scalebox{0.85}{
$P(Target, Feature) = (0,1)) = (1-p) \cdot q$
}
\end{center}

\\

\begin{center}
\scalebox{0.85}{
$P(Target, Feature) = (0,0)) = (1-p) \cdot (1-q)$
}
\end{center}

\\

Where $p$ and $q$ are fractions of positive observations in target and
feature respectively. An another view at permutation test is therefore that we 
get a contingency table, which is to be tested for independence.
Computing probability of a such table with two constraints, $n_{1,\cdot} = n_{1,1} + n_{1,0}$ and
$n_{\cdot, 1} = n_{1,1} + n_{0,1}$, and 
conditioning on $n_{1,1}$, leads to hypergeometric distribution.
$n_{i,j}$ denotes number of observations for which 
\scalebox{0.85}{$(Target, Feature) = (i,j)$}

This is in fact exact two-sided Fisher's test~\citep{lehmann1986testing}. 
%Information Gain is a way of deciding how strong is the evidence against independance.

\end{block}
\vfill 
    
    }
        \end{minipage}
      \end{beamercolorbox}
    \end{column}
    
    
%new column ------------------------------------------------------    
    
    \begin{column}{.54\textwidth}
      \begin{beamercolorbox}[center,wd=\textwidth]{postercolumn}
        \begin{minipage}[T]{.95\textwidth}  
          \parbox[t][\columnheight]{\textwidth}
            {
     

 
\begin{block}{Computational cost}

The cost of performing QuiPT is equal to computing a test statistic and 
probability of occurrence for $n_{1,1} + n_{0,1}$ contingency tables.

Suppose we consider 6-grams build on sequences of length 25 build of four 
characters. Then there are around 100,000 n-grams (features) to test. 
This means that for Benjamini-Hochberg procedure, we need to calculate 
p-values with accuracy of $0.05 \times 10^{-5}$. This requires at least 
2 million permutations. Each permutation, apart from reshuffling labels, 
requires computation of a test statistic. Since n-gram features are very sparse vectors, QuiPT needs to evaluate only few contingency tables.

The relative difference in speed between QuiPT and normal permutation tests depends on several factors, as a number of permutations and input data. For example, for simulation scheme presented below, QuiPT was on average 93 times faster than normal permutation test with $10^5$ permutations.

\end{block}
\vfill 
    
\begin{block}{Distribution of Information Gain for given contingency table}

\begin{columns}
\column{.04\textwidth}
\column{.35\textwidth}
\centering
Given constraint on $n_{1,1} + n_{0,1}$, probability distribution on
contingency tables, which permutations might produce, 
can be computed exactly.
\column{.65\textwidth}
\scalebox{0.79}{
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}

{\centering \includegraphics[width=\maxwidth]{figure/unnamed-chunk-5-1} 

}



\end{knitrout}
}
\end{columns}
\end{block}
\vfill

\begin{block}{Simulation scheme - genomics}

\begin{enumerate}[1.]
\item Random 4000 sequences (20 nucleotides each). The half of the sequences has label 0.
\item Choose a single position between 3 and 18 (to avoid border cases).
\item Resample nucleotides at chosen position. The dominant nucleotide has probability of occurrence $p_d = 0.25$. Other nucleotides have probability of occurrence $p_o = (1 - p_d)/3  $. 
\item Perform QuiPT (Information Gain as test statistic) and choose significant features (with p-value $< 0.001$).
\item Iterate steps 1-4 over other values of $p_d$ - 0.38, 0.51, 0.65, 0.78, 0.91.
\item Repeat steps 1-5 200 times.
\end{enumerate}
    
\end{block}
\vfill

\begin{block}{Power and False discoveries}
    \centering
\scalebox{0.9}{  
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}

{\centering \includegraphics[width=\maxwidth]{figure/unnamed-chunk-6-1} 

}



\end{knitrout}
}

    
    \end{block}
    \vfill
      
     
\begin{block}{Summary}

Quick permutation test is a powerful and quick equivalent of permutation test in a binary feature -- binary target testing scenario.
It is especially useful when precisely computed p-values are required and features are sparse vectors.
\end{block}
\vfill 
    
        \begin{block}{Avaibility}
        \footnotesize{
      
      QuiPT is a part of \textbf{biogram} R package devoted to the analysis of n-grams extracted from biological sequences: 
      
      \url{http://cran.r-project.org/web/packages/biogram/}
        }
    \end{block}
    \vfill 
     
     
    \begin{block}{Bibliography}
    \tiny{
      \bibliographystyle{apalike}
      \bibliography{biogram_poster}
    }
    \end{block}
    \vfill
            }
        \end{minipage}
      \end{beamercolorbox}
    \end{column}
  \end{columns}  
\end{frame}
\end{document}
